src/system/WatchableSocketSelect.cpp

```diff
@@ -0,0 +1,273 @@
+ *    Copyright (c) 2014-2017 Nest Labs, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements WatchableEvents using select().
+ */
+
+#include <platform/LockTracker.h>
+#include <support/CodeUtils.h>
+#include <system/SystemLayer.h>
+#include <system/SystemSockets.h>
+
+#include <errno.h>
+
+#define DEFAULT_MIN_SLEEP_PERIOD (60 * 60 * 24 * 30) // Month [sec]
+
+#if CHIP_DEVICE_CONFIG_ENABLE_MDNS && !__ZEPHYR__
+
+namespace chip {
+namespace Mdns {
+// TODO(#5556): Convert MDNS to WatchableSocket.
+void UpdateMdnsDataset(fd_set & readFdSet, fd_set & writeFdSet, fd_set & errorFdSet, int & maxFd, timeval & timeout);
+void ProcessMdns(fd_set & readFdSet, fd_set & writeFdSet, fd_set & errorFdSet);
+} // namespace Mdns
+} // namespace chip
+
+#endif // CHIP_DEVICE_CONFIG_ENABLE_MDNS && !__ZEPHYR__
+
+namespace chip {
+namespace System {
+
+void WatchableEventManager::Init(Layer & systemLayer)
+{
+    mSystemLayer = &systemLayer;
+    mMaxFd       = -1;
+    FD_ZERO(&mRequest.mReadSet);
+    FD_ZERO(&mRequest.mWriteSet);
+    FD_ZERO(&mRequest.mErrorSet);
+}
+
+void WatchableEventManager::Shutdown()
+{
+    mSystemLayer = nullptr;
+}
+
+/**
+ *  Set the read, write or exception bit flags for the specified socket based on its status in
+ *  the corresponding file descriptor sets.
+ *
+ *  @param[in]    socket    The file descriptor for which the bit flags are being set.
+ *
+ *  @param[in]    readfds   A pointer to the set of readable file descriptors.
+ *
+ *  @param[in]    writefds  A pointer to the set of writable file descriptors.
+ *
+ *  @param[in]    exceptfds  A pointer to the set of file descriptors with errors.
+ */
+SocketEvents WatchableEventManager::SocketEventsFromFDs(int socket, const fd_set & readfds, const fd_set & writefds,
+                                                        const fd_set & exceptfds)
+{
+    SocketEvents res;
+
+    if (socket >= 0)
+    {
+        // POSIX does not define the fd_set parameter of FD_ISSET() as const, even though it isn't modified.
+        if (FD_ISSET(socket, const_cast<fd_set *>(&readfds)))
+            res.Set(SocketEventFlags::kRead);
+        if (FD_ISSET(socket, const_cast<fd_set *>(&writefds)))
+            res.Set(SocketEventFlags::kWrite);
+        if (FD_ISSET(socket, const_cast<fd_set *>(&exceptfds)))
+            res.Set(SocketEventFlags::kError);
```

Comment:
It is a misconception that exceptfds is related to the occurrence of errors.

Reply:
True, that's a holdover from inherited code. I think `SocketEvents` can get a fourth flag for clarity at no additional cost.

Reply:
Ack